README.adoc

```diff
@@ -56,148 +56,97 @@ The driver will use default credentials set in the Google Cloud SDK `gcloud` app
-
-The lack of the foreign key constraint affects relationships and collection properties annotated with `@ElementCollection`.
-
-
-==== Generated IDs
-
-Hibernate's `@GeneratedValue` annotation for `long` properties is supported but not recommended:
 [source, java]
 ----
 @Entity
 public class Employee {
 
   @Id
-  @GeneratedValue // DON'T DO THIS, IT'S SLOW
-  public Long id;
+  @GeneratedValue(strategy = GenerationType.AUTO)
+  @Type(type="uuid-char")
+  public UUID id;
 }
 ----
 
-This results in sequential IDs that are https://cloud.google.com/spanner/docs/schema-and-data-model#primary_keys[not optimal for Cloud Spanner] and requires locking of the `hibernate_sequence` table for inserts.
+The `@Type(type="uuid-char")` annotation specifies that this UUID value will be stored in Cloud Spanner as a `STRING` column.
+Leaving out this annotation causes a `BYTES` column to be used.
 
-You should use locally-generated UUID key values instead:
+Hibernate's `@GeneratedValue` annotation for `long` properties is supported but not recommended:
 [source, java]
 ----
 @Entity
 public class Employee {
 
   @Id
-  @GeneratedValue(strategy = GenerationType.AUTO)
-  @Type(type="uuid-char")
-  public UUID id;
+  @GeneratedValue   // Not Recommended.
+  public Long id;
 }
 ----
 
-The `@Type(type="uuid-char")` annotation specifies that this UUID value will be stored in Cloud Spanner as a `STRING` column.
-Leaving out this annotation causes a `BYTES` column to be used.
+This results in sequential IDs that are https://cloud.google.com/spanner/docs/schema-and-data-model#primary_keys[not optimal for Cloud Spanner] and requires locking of the `hibernate_sequence` table for inserts.
 
+=== Limit Large DML Transactions
 
-==== Table Catalogs and Schemas
+Cloud Spanner implements a mutation limit on each transaction - each Spanner transaction https://cloud.google.com/spanner/quotas#limits_for_creating_reading_updating_and_deleting_data[may only have up to 20,000 operations which modify rows of a table].
 
-Cloud Spanner does not support table names with catalog and schema components:
+NOTE: Deleting a row counts as one operation and inserting/updating rows will count with the multiplicity of the number of columns they affect.
+For example if one inserts a row that contains 5 columns, it counts as 5 modify operations for the insert.
 
-[source, java]
-----
-// Not supported: `public.store.book` is not a valid Cloud Spanner table name reference.
-@Table(
-  catalog = "public",
-  schema = "store",
-  name = "book"
-)
+Consequently, users must take care to work around the limitations.
 
-// Supported.
-@Table(
-  name = "book"
-)
-----
+1. We recommend avoiding the usage of `CASCADE_TYPE.ALL` in Entity annotations because this might trigger a large number of entities to be deleted in a single transaction and bring you over the 20,000 limit.
```

Comment:
```suggestion
1. We recommend being careful with the use of `CASCADE_TYPE.ALL` in Entity annotations because, depending on the application, it might trigger a large number of entities to be deleted in a single transaction and bring you over the 20,000 limit.
```